File: Øving2.c
// gcc 4.7.2 +
// gcc -std=gnu99 -Wall -g -o oving2 oving2.c -lpthread

#include <pthread.h>
#include <stdio.h>
#include <stdint.h>
#include <time.h>

pthread_mutex_t lock;
int32_t x = 0, y = 0;

void* fast_thread1()
{
    pthread_mutex_lock(&lock);

    for (uint32_t j = 0; j < 1000000; j++) {
    	x++;
    }
    
    pthread_mutex_unlock(&lock);

    return NULL;
}

void* fast_thread2()
{
    pthread_mutex_lock(&lock);

    for (uint32_t j = 0; j < 1000000 - 1; j++) {
    	x--;
    }
    
    pthread_mutex_unlock(&lock);

    return NULL;
}

void* slow_thread1()
{
    for (uint32_t j = 0; j < 1000000; j++) {
        pthread_mutex_lock(&lock);
    	y++;
        pthread_mutex_unlock(&lock);
    }

    return NULL;
}

void* slow_thread2()
{
    for (uint32_t j = 0; j < 1000000 - 1; j++) {
        pthread_mutex_lock(&lock);
    	y--;
        pthread_mutex_unlock(&lock);
    }

    return NULL;
}



int main()
{
    clock_t start, diff;
    int msec;
    pthread_t ft1, ft2, st1, st2;
    pthread_mutex_init(&lock, NULL);
    
    // Fast locking
    start = clock();
    pthread_create(&ft1, NULL, fast_thread1, NULL);
    pthread_create(&ft2, NULL, fast_thread2, NULL);
    
    pthread_join(ft1, NULL);
    pthread_join(ft2, NULL);
    diff = clock() - start;
    msec = diff * 1000 / CLOCKS_PER_SEC;
    printf("fast x = %d, took %d milliseconds!\n", x, msec);
    
    // Slow locking
    start = clock();
    pthread_create(&st1, NULL, slow_thread1, NULL);
    pthread_create(&st2, NULL, slow_thread2, NULL);
    
    pthread_join(st1, NULL);
    pthread_join(st2, NULL);
    diff = clock() - start;
    msec = diff * 1000 / CLOCKS_PER_SEC;
    printf("slow y = %d, took %d milliseconds!\n", y, msec);
    
    pthread_mutex_destroy(&lock);
    return 0;
    
}

File: Øving2.py
# Python 3.3.3 and 2.7.6
# python oving2.py

from threading import Thread, Lock
from time import time

x, y = 0, 0
lock = Lock()

def fast_thread1():
    global x
    with lock:
        for _ in range(1000000):
            x += 1

def fast_thread2():
    global x
    with lock:
        for _ in range(1000000 - 1):
            x -= 1  
    
def slow_thread1():
    global y
    for _ in range(1000000):
        with lock:
            y += 1
        
def slow_thread2():
    global y
    for _ in range(1000000 - 1):
        with lock:
            y -= 1

def main():
    ft1 = Thread(target=fast_thread1, args=())
    ft2 = Thread(target=fast_thread2, args=())
    st1 = Thread(target=slow_thread1, args=())
    st2 = Thread(target=slow_thread2, args=())
    
    start = time()
    ft1.start()
    ft2.start()
    
    ft1.join()
    ft2.join()
    diff = time() - start
    print("fast x = %i, took %0.3f ms" % (x, diff*1000.0))
    
    start = time()
    st1.start()
    st2.start()
    
    st1.join()
    st2.join()
    diff = time() - start
    print("slow x = %i, took %0.3f ms" % (y, diff*1000.0))


main()
